### PR TITLE
Add complex_to_float option in ReductionOpInfo

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #78350
* #78424
* __->__ #78423

As per title

### CODE REVIEW

File: torch/testing/_internal/common_methods_invocations.py [hunk 1]
```diff
@@ -1381,6 +1378,7 @@ def __init__(
         # These are mutually exclusive options
         assert not (result_dtype and promotes_int_to_float)
         assert not (result_dtype and promotes_int_to_int64)
+        assert not (result_dtype and complex_to_real)
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:1381]: but `vector_norm` accepts `dtype` arg?

**Review by ngimel — COMMENTED** [review]
This looks good, but what's up with test failures?

File: torch/testing/_internal/common_methods_invocations.py (hunk 1 above)
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:1381]: `result_dtype` is used to specify when a function always returns a given dtype (e.g. `eq`).

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -1129,17 +1129,11 @@ def sample_inputs_reduction(op_info, device, dtype, requires_grad, **kwargs):
...
         for reduction_kwargs in _generate_reduction_kwargs(t.ndim, supports_multiple_dims):
             for args, kwargs in generate_args_kwargs(t, **reduction_kwargs):
                 kwargs.update(reduction_kwargs)
-                inputs.append(SampleInput(
-                    t.clone().requires_grad_(requires_grad),
-                    args=args,
-                    kwargs=kwargs))
-
-    return inputs
+                yield SampleInput(t.detach().requires_grad_(requires_grad), args=args, kwargs=kwargs)
```
[the remark this answers is not in the document]
Comment (Lezcano) [reply torch/testing/_internal/common_methods_invocations.py:1136]: I do not understand how the previous implementation worked tbh... I should look further into that.

This little fix needed some extra fixes here and there (hence the test failures). I needed to convert quite a few other functions into generators after this one. Let's see if I managed to not miss any one.

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**